chainlink-modsec/libmodsec/internal/executor/main.go

```diff
@@ -0,0 +1,117 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+func (e *Executor) run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var wg sync.WaitGroup
+
+	// 1. We read messages from all sources for any new incoming work.
+	// Immediately we perform checks to see if it's our turn to execute the message, and skip additional processing if not.
+	for _, reader := range e.MessageSources {
+		go func(r MessageReader) {
+			wg.Add(1)
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					fmt.Println("Executor shutting down message reader...")
+					return
+				case msg := <-reader.SubscribeMessages(ctx):
```

Comment:
This calls `SubscribeMessages` every time the branch is exercised? I'm not really sure how this function will be implemented but presumably a new channel will be created for each subscriber, perhaps better to call once and listen to the chan?